### PR TITLE
[FEATURE] `NumericMetricRangeMultiBatchParameterBuilder` kernel density estimation

### PR DESCRIPTION
**Pull request by NathanFarmer**
Changes proposed in this pull request:
- GREAT-706
- Add kernel density estimation (KDE) method to `NumericMetricRangeMultiBatchParameterBuilder`
- Apply consistent naming conventions across methods that share parameters and cannot be used together

### Definition of Done
- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [X] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 1]
```diff
@@ -98,35 +109,40 @@ def __init__(
     ) -> None:
         """
         Args:
-            name: the name of this parameter -- this is user-specified parameter name (from configuration);
-            it is not the fully-qualified parameter name; a fully-qualified parameter name must start with "$parameter."
-            and may contain one or more subsequent parts (e.g., "$parameter.<my_param_from_config>.<metric_name>").
+            name: the name of this parameter -- this is user-specified parameter name (from configuration); it is not
+                the fully-qualified parameter name; a fully-qualified parameter name must start with "$parameter."
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:113]: @NathanFarmer what is being changed here?  Just the indentation?  Do we need it?  Thanks.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 2]
```diff
@@ -98,35 +109,40 @@ def __init__(
...
+                and may contain one or more subsequent parts (e.g., "$parameter.<my_param_from_config>.<metric_name>").
             metric_name: the name of a metric used in MetricConfiguration (must be a supported and registered metric)
             metric_domain_kwargs: used in MetricConfiguration
             metric_value_kwargs: used in MetricConfiguration
             enforce_numeric_metric: used in MetricConfiguration to insure that metric computations return numeric values
             replace_nan_with_zero: if False, then if the computed metric gives NaN, then exception is raised; otherwise,
-            if True (default), then if the computed metric gives NaN, then it is converted to the 0.0 (float) value.
+                if True (default), then if the computed metric gives NaN, then it is converted to the 0.0 (float) value.
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:120]: @NathanFarmer Is this just the indentation?

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 3]
```diff
@@ -98,35 +109,40 @@ def __init__(
...
             reduce_scalar_metric: if True (default), then reduces computation of 1-dimensional metric to scalar value.
             false_positive_rate: user-configured fraction between 0 and 1 expressing desired false positive rate for
-            identifying unexpected values as judged by the upper- and lower- quantiles of the observed metric data.
+                identifying unexpected values as judged by the upper- and lower- quantiles of the observed metric data.
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:123]: @NathanFarmer Is this just the indentation?

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 4]
```diff
@@ -98,35 +109,40 @@ def __init__(
...
+            bw_method: Applicable only for the "kde" sampling method -- if omitted (default), then "scott" is used.
+                Possible values for the estimator bandwidth method are described at:
+                https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.gaussian_kde.html
+            random_seed: Applicable only for the "bootstrap" and "kde" sampling methods -- if omitted (default), then
+                uses "np.random.choice"; otherwise, utilizes "np.random.Generator(np.random.PCG64(random_seed))".
+            include_estimator_samples_histogram_in_details: Applicable only for the "bootstrap" sampling method -- if
+                True, then add 10-bin histogram of bootstraps to "details"; otherwise, omit this information (default).
             truncate_values: user-configured directive for whether or not to allow the computed parameter values
-            (i.e., lower_bound, upper_bound) to take on values outside the specified bounds when packaged on output.
+                (i.e., lower_bound, upper_bound) to take on values outside the specified bounds when packaged on output.
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:139]: @NathanFarmer Is this just the indentation?

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 5]
```diff
@@ -98,35 +109,40 @@ def __init__(
...
             round_decimals: user-configured non-negative integer indicating the number of decimals of the
-            rounding precision of the computed parameter values (i.e., min_value, max_value) prior to packaging them on
-            output.  If omitted, then no rounding is performed, unless the computed value is already an integer.
+                rounding precision of the computed parameter values (i.e., min_value, max_value) prior to packaging them
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:141]: @NathanFarmer Is this just the indentation?

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 6]
```diff
@@ -98,35 +109,40 @@ def __init__(
...
+                on output.  If omitted, then no rounding is performed, unless the computed value is already an integer.
             evaluation_parameter_builder_configs: ParameterBuilder configurations, executing and making whose respective
-            ParameterBuilder objects' outputs available (as fully-qualified parameter names) is pre-requisite.
-            These "ParameterBuilder" configurations help build parameters needed for this "ParameterBuilder".
+                ParameterBuilder objects' outputs available (as fully-qualified parameter names) is pre-requisite.
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:144]: @NathanFarmer Is this just the indentation?

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 7]
```diff
@@ -98,35 +109,40 @@ def __init__(
...
+            include_estimator_samples_histogram_in_details: Applicable only for the "bootstrap" sampling method -- if
+                True, then add 10-bin histogram of bootstraps to "details"; otherwise, omit this information (default).
             truncate_values: user-configured directive for whether or not to allow the computed parameter values
-            (i.e., lower_bound, upper_bound) to take on values outside the specified bounds when packaged on output.
+                (i.e., lower_bound, upper_bound) to take on values outside the specified bounds when packaged on output.
             round_decimals: user-configured non-negative integer indicating the number of decimals of the
-            rounding precision of the computed parameter values (i.e., min_value, max_value) prior to packaging them on
-            output.  If omitted, then no rounding is performed, unless the computed value is already an integer.
+                rounding precision of the computed parameter values (i.e., min_value, max_value) prior to packaging them
+                on output.  If omitted, then no rounding is performed, unless the computed value is already an integer.
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:142]: @NathanFarmer Is this just the indentation?

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 8]
```diff
@@ -665,9 +691,68 @@ def _get_bootstrap_estimate(
         return compute_bootstrap_quantiles_point_estimate(
             metric_values=metric_values,
             false_positive_rate=false_positive_rate,
+            quantile_statistic_interpolation_method=quantile_statistic_interpolation_method,
             n_resamples=n_resamples,
             random_seed=random_seed,
+        )
+
+    @staticmethod
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:699]: ❤️

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 9]
```diff
@@ -665,9 +691,68 @@ def _get_bootstrap_estimate(
...
+            parameters=parameters,
+        )
+
+        if n_resamples is None:
+            n_resamples = (
+                NumericMetricRangeMultiBatchParameterBuilder.DEFAULT_KDE_NUM_RESAMPLES
+            )
+
+        # Obtain bw_method override from "rule state" (i.e., variables and parameters); from instance variable otherwise.
+        bw_method: Optional[Any] = get_parameter_value_and_validate_return_type(
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: @NathanFarmer Is the return type of `Any` the best we can expect here?  The property returns something else -- perhaps a typo?  Thanks.

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@NathanFarmer Some questions and a potential small typo.  Thanks -- amazing!

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 1 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:113]: @alexsherstinsky I find it very difficult to read without the indentation.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 2 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:120]: @alexsherstinsky I find it very difficult to read without the indentation.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 3 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:123]: @alexsherstinsky I find it very difficult to read without the indentation.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 4 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:139]: @alexsherstinsky I find it very difficult to read without the indentation.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 5 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:141]: @alexsherstinsky I find it very difficult to read without the indentation.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 6 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:144]: @alexsherstinsky I find it very difficult to read without the indentation.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 7 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:142]: @alexsherstinsky I find it very difficult to read without the indentation.

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 9 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: @alexsherstinsky This parameter accepts quite a few types. I will update this to `Optional[Union[str, float, Callable]]`

Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py]: I linked to the acceptable parameter values in the `NumericMetricRangeMultiBatchParameterBuilder` docstring (https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.gaussian_kde.html) and I updated this type hint everywhere. ✅

**Review by alexsherstinsky — APPROVED** [review]
LGTM

**Pull request merged**